app/models/assignment.rb

```diff
@@ -838,4 +892,106 @@ def update_assigned_tokens
       t.update_tokens(tokens_per_day_was, tokens_per_day)
     end
   end
+
+  def add_new_grouping_for_group(row, group)
+    # Create a new Grouping for this assignment and the newly
+    # crafted group
+    grouping = Grouping.new(assignment: self, group: group)
+    grouping.save
+
+    # Form groups
+    start_index_group_members = 2
+    (start_index_group_members..(row.length - 1)).each do |i|
+      student = Student.where(user_name: row[i])
+                    .first
+      if student
+        if grouping.student_membership_number == 0
+          # Add first valid member as inviter to group.
+          grouping.group_id = group.id
+          grouping.save # grouping has to be saved, before we can add members
+
+          # We could call grouping.add_member, but it updates repo permissions
+          # For performance reasons in the csv upload we will just create the
+          # member here, and do the permissions update as a bulk operation.
+          member = StudentMembership.new(
+            user: student,
+            membership_status: StudentMembership::STATUSES[:inviter],
+            grouping: grouping)
+          member.save
+        else
+          member = StudentMembership.new(
+            user: student,
+            membership_status: StudentMembership::STATUSES[:accepted],
+            grouping: grouping)
+          member.save
+        end
+      end
+    end
+  end
+
+  #
+  # Return true if for each membership given, a corresponding student exists
+  # and if they are not part of a different grouping for the same assignment
+  #
+  def membership_unique?(row)
+    start_index_group_members = 2 # index where student names start in the row
+    (start_index_group_members..(row.length - 1)).each do |i|
+      student = Student.find_by user_name: row[i]
+      if student
+        unless student.accepted_grouping_for(id).nil?
+          errors.add(:groupings, student.user_name)
+          return false
+        end
+      else
+        errors.add(:student_memberships, row[i])
+        return false
+      end
+    end
+    true
+  end
+
+  # Return true if the given membership in the csv row is the exact same as the
+  # membership of the given existing_grouping.
+  def same_membership_as_csv_row?(row, existing_grouping)
+    start_index_group_members = 2 # index where student names start in the row
+    # check if all the members given in the csv file exists and belongs to the
+    # given grouping
+    (start_index_group_members..(row.length - 1)).each do |i|
+      student = Student.find_by user_name: row[i]
+      if student
+        grouping = student
+            .accepted_grouping_for(existing_grouping.assignment.id)
```

Comment:
Place the . on the previous line, together with the method call receiver.<br>Align the operands of an expression in an assignment spanning multiple lines.